rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java

```diff
@@ -3502,15 +3547,236 @@ private Object js_SymbolMatchAll(
         boolean fullUnicode = flags.indexOf('u') != -1 || flags.indexOf('v') != -1;
 
         return new NativeRegExpStringIterator(scope, matcher, s, global, fullUnicode);
     }
 
+    private Object js_SymbolReplace(
+            Context cx, Scriptable scope, Scriptable thisObj, Object[] args) {
+        // See ECMAScript spec 22.2.6.11
+        if (!ScriptRuntime.isObject(thisObj)) {
+            throw ScriptRuntime.typeErrorById("msg.arg.not.object", ScriptRuntime.typeof(thisObj));
+        }
+
+        String s = ScriptRuntime.toString(args.length > 0 ? args[0] : Undefined.instance);
+        int lengthS = s.length();
+        Object replaceValue = args.length > 1 ? args[1] : Undefined.instance;
+        boolean functionalReplace = replaceValue instanceof Callable;
+        if (!functionalReplace) {
+            replaceValue = ScriptRuntime.toString(replaceValue);
+        }
+        String flags = ScriptRuntime.toString(ScriptRuntime.getObjectProp(thisObj, "flags", cx));
+        boolean global = flags.indexOf('g') != -1;
+        boolean fullUnicode = flags.indexOf('u') != -1 || flags.indexOf('v') != -1;
+        if (global) {
+            setLastIndex(thisObj, ScriptRuntime.zeroObj);
+        }
+
+        List<Object> results = new ArrayList<>();
+        boolean done = false;
+        while (!done) {
+            Object result = regExpExec(thisObj, s, cx, scope);
+            if (result == null) {
+                done = true;
+            } else {
+                results.add(result);
+                if (!global) {
+                    done = true;
+                } else {
+                    String matchStr =
+                            ScriptRuntime.toString(
+                                    ScriptRuntime.getObjectIndex(result, 0, cx, scope));
+                    if (matchStr.isEmpty()) {
+                        long thisIndex = getLastIndex(cx, thisObj);
+                        long nextIndex =
+                                ScriptRuntime.advanceStringIndex(s, thisIndex, fullUnicode);
+                        setLastIndex(thisObj, nextIndex);
+                    }
+                }
+            }
+        }
+
+        StringBuilder accumulatedResult = new StringBuilder();
+        int nextSourcePosition = 0;
+        for (Object result : results) {
+            long resultLength =
+                    ScriptRuntime.toLength(
+                            ScriptRuntime.getObjectProp(result, "length", cx, scope));
+            long nCaptures = Math.max(resultLength - 1, 0);
+            String matched =
+                    ScriptRuntime.toString(ScriptRuntime.getObjectIndex(result, 0, cx, scope));
+            int matchLength = matched.length();
+            double positionDbl =
+                    ScriptRuntime.toInteger(
+                            ScriptRuntime.getObjectProp(result, "index", cx, scope));
+            int position;
+            if (positionDbl < 0) {
```

Comment:
This pattern (clamping the position to the length in some way) appears more than once in this code, and looking at the code coverage, we only test one of the three possible paths.